examples/demo-nft-module/src/call.rs

```diff
@@ -2,94 +2,314 @@ use anyhow::{bail, Result};
+
+        if let Some(mut c) = self.collections.get(&collection_address, working_set) {
+            let nft_identifier = NftIdentifier(token_id, collection_address.clone());
+            if c.frozen {
+                bail!(
+                    "Collection with name {} by sender {} is already frozen",
+                    collection_name,
+                    creator.to_string()
+                )
+            } else if self.nfts.get(&nft_identifier, working_set).is_some() {
+                bail!(
+                    "NFT id {} in Collection with name {}, creator {} already exists",
+                    token_id,
+                    collection_name,
+                    creator.to_string()
+                );
+            } else {
+                let new_nft = Nft {
+                    token_id,
+                    collection_address: collection_address.clone(),
+                    owner: mint_to_address.clone(),
+                    frozen,
+                    token_uri: collection_uri.to_string(),
+                };
+                self.nfts.set(&nft_identifier, &new_nft, working_set);
+                c.supply += 1;
+                self.collections.set(&collection_address, &c, working_set);
             }
-            Some(owner) => owner,
-        };
-        if &token_owner != context.sender() {
-            bail!("Only token owner can transfer token");
+        } else {
+            bail!(
+                "Collection with name {} by sender {} does not exist",
+                collection_name,
+                creator.to_string()
+            );
         }
-        self.owners.set(&id, &to, working_set);
-        working_set.add_event(
-            "NFT transfer",
-            &format!("A token with id {id} was transferred"),
-        );
+
         Ok(CallResponse::default())
     }
 
-    pub(crate) fn burn(
+    pub(crate) fn transfer_nft(
         &self,
-        id: u64,
+        nft_id: u64,
+        collection_address: &CollectionAddress<C>,
+        to: &UserAddress<C>,
         context: &C,
         working_set: &mut WorkingSet<C::Storage>,
     ) -> Result<CallResponse> {
-        let token_owner = match self.owners.get(&id, working_set) {
-            None => {
-                bail!("Token with id {} does not exist", id);
+        let owner = context.sender();
+
+        if self
+            .collections
+            .get(collection_address, working_set)
+            .is_some()
+        {
+            let nft_identifier = NftIdentifier(nft_id, collection_address.clone());
+            if let Some(mut n) = self.nfts.get(&nft_identifier, working_set) {
+                if owner.as_ref() == n.owner.as_ref() {
+                    n.owner = to.clone();
+                    self.nfts.set(&nft_identifier, &n, working_set);
+                } else {
+                    bail!("Transfer sent with owner {}, NFT id {} in Collection with address {} is owned by {}",
```

Comment:
This branch and the one in line 250 is repeated a few times.
What about adding methods like:
1. `check_owner(..)->Result`
2. `check_if_collection_exists()-> Result`
3. etc..

And the at the beginning of the methods that need validation we do:
```
check_owner()?;
check_if_collection_exists()?;
...
```

Reply:
yup that makes sense and it also keeps the core business logic separated and free from all the validation logic - avoids bugs too! will make this change